com.unity.netcode.gameobjects/Runtime/Serialization/BitReader.cs

```diff
@@ -0,0 +1,218 @@
+using System;
+using System.Runtime.CompilerServices;
+using Unity.Netcode;
+
+namespace Unity.Multiplayer.Netcode
+{
+    /// <summary>
+    /// Helper class for doing bitwise reads for a FastBufferReader.
+    /// Ensures all bitwise reads end on proper byte alignment so FastBufferReader doesn't have to be concerned
+    /// with misaligned reads.
+    /// </summary>
+    public ref struct BitReader
+    {
+        private Ref<FastBufferReader> m_Reader;
+        private readonly unsafe byte* m_BufferPointer;
+        private readonly int m_Position;
+        private int m_BitPosition;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+        private int m_AllowedBitwiseReadMark;
+#endif
+
+        private const int k_BitsPerByte = 8;
+
+        /// <summary>
+        /// Whether or not the current BitPosition is evenly divisible by 8. I.e. whether or not the BitPosition is at a byte boundary.
+        /// </summary>
+        public bool BitAligned
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => (m_BitPosition & 7) == 0;
+        }
+
+        internal unsafe BitReader(ref FastBufferReader reader)
+        {
+            m_Reader = new Ref<FastBufferReader>(ref reader);
+
+            m_BufferPointer = m_Reader.Value.BufferPointer + m_Reader.Value.Position;
+            m_Position = m_Reader.Value.Position;
+            m_BitPosition = 0;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_AllowedBitwiseReadMark = (m_Reader.Value.AllowedReadMark - m_Position) * k_BitsPerByte;
+#endif
+        }
+
+        /// <summary>
+        /// Pads the read bit count to byte alignment and commits the read back to the reader
+        /// </summary>
+        public void Dispose()
```

Comment:
Did you mean to implement IDisposable?

Reply:
ref structs can't implement interfaces, but c# has special behavior to support `using()` on ref structs that have a `public void Dispose()`

Reply:
Ohh, neat. ref structs are pretty new to me. I'm a little uneasy about some of your changes in general because I haven't dipped into some of the new ref syntax - especially the Ref<T> type you're adding. I hope you're seeking review from some folks that have a deep experience with this since this type of code is very risky

Reply:
(To be clear, not saying you shouldn't be doing it, serialization is exactly where I'd expect to be writing this type of code)